mypy/checker.py

```diff
@@ -280,38 +320,57 @@ def get_generator_yield_type(self, return_type: Type) -> Type:
             return AnyType()
-        elif not self.is_generator_return_type(return_type):
+        elif not self.is_generator_return_type(return_type, is_coroutine):
             # If the function doesn't have a proper Generator (or superclass) return type, anything
             # is permissible.
             return AnyType()
         elif not isinstance(return_type, Instance):
             # Same as above, but written as a separate branch so the typechecker can understand.
             return AnyType()
         elif return_type.type.fullname() == 'typing.Generator':
-            # Generator is the only type which specifies the type of values it can receive.
-            return return_type.args[1]
+            # Generator is one of the two types which specify the type of values it can receive.
+            if len(return_type.args) == 3:
+                return return_type.args[1]
+            else:
+                return AnyType()
+        elif return_type.type.fullname() == 'typing.Awaitable':
+            # Generator is one of the two types which specify the type of values it can receive.
+            # According to the stub this is always `Any`.
+            return AnyType()
         else:
             # `return_type` is a supertype of Generator, so callers won't be able to send it
             # values.
             return Void()
 
-    def get_generator_return_type(self, return_type: Type) -> Type:
+    def get_generator_return_type(self, return_type: Type, is_coroutine: bool) -> Type:
+        """Given the declared return type of a generator (t), return the type it returns (tr)."""
         if isinstance(return_type, AnyType):
             return AnyType()
-        elif not self.is_generator_return_type(return_type):
+        elif not self.is_generator_return_type(return_type, is_coroutine):
             # If the function doesn't have a proper Generator (or superclass) return type, anything
             # is permissible.
             return AnyType()
         elif not isinstance(return_type, Instance):
             # Same as above, but written as a separate branch so the typechecker can understand.
             return AnyType()
         elif return_type.type.fullname() == 'typing.Generator':
-            # Generator is the only type which specifies the type of values it returns into
-            # `yield from` expressions.
-            return return_type.args[2]
+            # Generator is one of the two types which specify the type of values it returns into
+            # `yield from` expressions (using a `return` statements).
+            if len(return_type.args) == 3:
+                return return_type.args[2]
+            else:
+                return AnyType()
+        elif return_type.type.fullname() == 'typing.Awaitable':
+            # Awaitable is the other type which specifies the type of values it returns into
+            # `yield from` expressions (using `return`).
+            if len(return_type.args) == 1:
```

Comment:
IIUC when you use a generic class without parameters args is empty. I also worry about crashes when someone writes e.g. `Generator[int]` -- it may well produce an error message but also an Instance with `len(args) == 1`, so I'm doing this defensively.